### DISCUSSION TITLE
Incorrect behaviour after M0 in gcode

### DISCUSSION THREAD

**Discussion by sdmacleod** [bigtreetech/BIGTREETECH-TouchScreenFirmware#2228] [Q&A]
So I used Cura's "Pause at height" gcode script to insert a pause so that I can put a magnet in a print.  The script moves the head out of the way of the print (back corner).
But after touching "Resume" on the TFT35v2, the extruder retracts (a lot) and moves to (0,0,0), before moving to the last print location.
I've traced the issue as far as the "PrintPause" function in Printing.c - past line 582, we're storing tmp.axis - but it appears that no location has ever been initialized into tmp.axis.

Never been on Github for this stuff- do I create an New Issue for this?

**Comment by kisslorand** [comment]
tmp.axis is initialised in Printing.c at line 550 by `coordinateGetAll(&tmp)`.
Anyway the continue after M0 pause procedure is handled by Marlin by M108, not the TFT FW. In case of M0 pause, after resume, the TFT sends M108 by `breakAndContinue()` and than exits by `break` (line 580) so no X, Y, Z, E restore action is taken by the TFT.

**Reply by sdmacleod** [reply]
I agree that that's what I would expect, but that's not the behaviour I'm observing.
Under "else  // resume" (line 581), the if statement "if (infoPrinting.pauseType < PAUSE_EXTERNAL)" engages (because the gcode called the M0, not the TFT).  The lines past that inject G1 moves to the axis locations the TFT recorded just prior to the pause (as you say, at line 550). But nothing is getting saved to tmp.axis
I've commented out lines 586 to 608 inclusive, and replaced them with:
          breakAndContinue();  // clear the queue and send a break and continue
          break;
And now it works as expected.  I don't understand why the TFT code is trying

I have a hard time believing nobody else would ever have seen this, so I'd love to figure it out.  And I'd be interested in someone else's run.

If you'd care to try it, this gcode pauses after the first layer.  If moves to (0,0,0) after resuming from the pause, this is issue I'm seeing.  If not, its just me.  I believe this will only occur in non-Marlin mode (can't test Marlin mode on my v2)
[test_cylinder.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/7422050/test_cylinder.zip)

**Reply by kisslorand** [reply]
> I agree that that's what I would expect, but that's not the behaviour I'm observing. Under "else // resume" (line 581), the if statement "if (infoPrinting.pauseType < PAUSE_EXTERNAL)" engages (because the gcode called the M0, not the TFT). 

"interfaceCmd.c" parses the G-codes read by the TFT. In "interfaceCmd.c" line 312 "printPause(true, PAUSE_M0)" is called which sets "infoPrinting.pauseType" to "PAUSE_M0".  I do not know how it is possible that "if (infoPrinting.pauseType < PAUSE_EXTERNAL)" engages...

**Comment by radek8** [comment]
I tried your test file.
The behavior is as you describe. After confirming the pause and clicking continue, the nozzle moves to the front left corner and touches the pad (pushes approx. 0.5 mm into the pad) Grrr! .

**Reply by sdmacleod** [reply]
Oh no!  I hope your bed is okay?  I can't think why it would drop so far.  Mine just goes to zero...  Anyway, thanks for trying it.

I haven't had time to reinvestigate since my initial posts.  And the printer is in pieces at the moment...
